### ISSUE TITLE
sql: review slow tests

### ISSUE THREAD

**Issue by RaduBerinde** [cockroachdb/cockroach#37028]
The SQL tests have been getting very slow. Here are a list of the slowest tests in the `sql` package. We should review these tests and see if we can trim them down or speed them up. We should also add `t.Parallel()` where appropriate.

We can also make some tests run fewer iterations on non-nightly runs.

 - [ ] TestScatterRandomizeLeases (22.90s)
 - [ ] TestRaceWithBackfill (6.68s)
 - [ ] TestCancelSchemaChange (6.00s)
 - [ ] TestDistBackfill (5.83s)
 - [ ] TestBackfillCompletesOnChunkBoundary (5.81s)
 - [ ] TestBackfillErrors (5.44s)
 - [ ] TestDropWhileBackfill (5.38s)
 - [ ] TestDropTableDeleteData (4.89s)
 - [ ] TestSchemaChangeEvalContext (3.04s)
 - [x] TestQueryCache ~~(2.89s)~~ Now: 0.5s
 - [ ] TestDropAndCreateDatabase (2.59s)
 - [ ] TestDropAndCreateTable (2.41s)
 - [ ] TestScatterResponse (2.39s)
 - [ ] TestDropTableWhileUpgradingFormat (2.32s)
 - [ ] TestCancelSelectQuery (2.21s)
 - [ ] TestDropDatabaseDeleteData (1.86s)
 - [ ] TestSchemaChangeReverseMutations (1.52s)
 - [ ] TestShowCreateTable (1.47s)
 - [ ] TestTrace (1.32s)
 - [ ] TestCancelSchemaChangeContext (1.29s)
 - [ ] TestSchemaChangeRetry (1.26s)
 - [ ] TestActiveCancelSession (1.26s)
 - [ ] TestSchemaChangeRetryOnVersionChange (1.23s)
 - [ ] TestMultipleIndexBackfills (1.13s)
 - [ ] TestFuncNull (1.04s)
 - [ ] TestTruncateCompletion (1.03s)
 - [ ] TestAsyncSchemaChanger (1.01s)

**Comment by jordanlewis** [comment]
About 90% of these are bulkio/schema change tests, 5% are core (TestScatterRandomize), and 5% are sql, just for the record.

**Comment by jordanlewis** [comment]
And `TestQueryCache` is opt.

**Comment by RaduBerinde** [comment]
Note: when reviewing these, we should look at the running time in relation to what it is testing. For example, I think 1.5s for testing `SHOW CREATE TABLE` is excessive while 5s for a schema change test might be reasonable.

**Comment by RaduBerinde** [comment]
@andy-kimball can someone from core look at `TestScatterRandomizeLeases`?

**Comment by jordanlewis** [comment]
Bumping this as many CI flakes are now being caused by timing out test jobs.

Here's the sorted list of all packages that take longer than 10 seconds to run their tests:

```
ok  	github.com/cockroachdb/cockroach/pkg/storage/tscache	11.665s
ok  	github.com/cockroachdb/cockroach/pkg/sql/pgwire	13.859s
ok  	github.com/cockroachdb/cockroach/pkg/sql/stats	16.653s
ok  	github.com/cockroachdb/cockroach/pkg/storage/engine	19.744s
ok  	github.com/cockroachdb/cockroach/pkg/ts	19.877s
ok  	github.com/cockroachdb/cockroach/pkg/sql/sem/builtins	21.239s
ok  	github.com/cockroachdb/cockroach/pkg/sql/opt/exec/execbuilder	24.905s
ok  	github.com/cockroachdb/cockroach/pkg/workload/ycsb	25.039s
ok  	github.com/cockroachdb/cockroach/pkg/kv	25.621s
ok  	github.com/cockroachdb/cockroach/pkg/storage/spanlatch	25.936s
ok  	github.com/cockroachdb/cockroach/pkg/sql/tests	30.790s
ok  	github.com/cockroachdb/cockroach/pkg/cli	32.324s
ok  	github.com/cockroachdb/cockroach/pkg/ccl/workloadccl/allccl	43.270s
ok  	github.com/cockroachdb/cockroach/pkg/ccl/backupccl	47.053s
ok  	github.com/cockroachdb/cockroach/pkg/server	57.771s
ok  	github.com/cockroachdb/cockroach/pkg/ccl/importccl	68.770s
ok  	github.com/cockroachdb/cockroach/pkg/ccl/changefeedccl	73.780s
ok  	github.com/cockroachdb/cockroach/pkg/storage	193.096s
ok  	github.com/cockroachdb/cockroach/pkg/sql	201.081s
ok  	github.com/cockroachdb/cockroach/pkg/ccl/partitionccl	230.300s
```

**Comment by jordanlewis** [comment]
Sorted list of individual tests that take longer than 10 seconds:

```
--- PASS: TestInvertedIndex 10.07s
--- PASS: TestRocksDBDeleteRangeCompaction 10.39s
--- PASS: TestWedgedReplicaDetection 10.45s
--- PASS: TestMonotonicInserts 11.68s
--- PASS: TestZetaIncrementally 11.82s
--- PASS: TestChangefeedNemeses 12.21s
--- PASS: TestZetaFromScratch 13.14s
--- PASS: TestImportCSVStmt 16.11s
--- PASS: TestAllRegisteredImportFixture 18.58s
--- PASS: TestSlidingWindow 20.80s
--- PASS: TestImportData 21.89s
--- PASS: TestAllRegisteredSetup 24.44s
--- PASS: TestExecBuild 24.51s
--- PASS: TestBTree 24.56s
--- PASS: TestScatterRandomizeLeases 28.23s
--- PASS: TestMergeQueue 28.48s
--- PASS: TestInitialPartitioning 99.15s
--- PASS: TestRepartitioning 127.83s
```

**Comment by andy-kimball** [comment]
The Repartitioning tests are not really single tests, but have many subtests. If we broke the subtests up into separate tests, they wouldn't show up as being "slow". But I'm assuming that wouldn't be helpful. Aren't the timeouts caused by the entire test suite running too long, rather than by individual tests running too long? If it's the entire test suite, then won't we inevitably hit whatever arbitrary time limit we set eventually, since we're continually adding new tests every release? Shouldn't we instead look at running the tests in parallel on more CI machines?

**Comment by RaduBerinde** [comment]
TestInitialPartitioning and TestRepartitioning are running a lot of testcases on the same server. I would consider breaking it up into a few parallel subtests that can be run in parallel. Or at least call `t.Parallel()` so the two run in parallel with each other.

**Comment by jordanlewis** [comment]
I tried running those in parallel by making multiple servers, but that didn't make them any faster.

I think we could try changing the tests so that they use different databases and running those in parallel, but not sure how much mileage that will get us either.

Ultimately I agree with Andy that over time we'll need to scale out our testing. But for now I do think we could make some improvements case by case.

**Comment by RaduBerinde** [comment]
This patch takes down TestInitialPartitioning from 99s to 20s for me: https://gist.github.com/RaduBerinde/051e83d5d8863b6e3bc91379356e74ec

**Comment by RaduBerinde** [comment]
Scaling out CI testing is perpendicular to making sure the tests we have are not wasteful. We found many instances where tests were running way too many iterations or weren't using parallelism, and I'm sure there are many more. These should be addressed regardless - for one they are experienced daily by everyone when `make test` is run locally.

**Comment by andy-kimball** [comment]
Doesn't your patch just make that single test run faster in isolation? That is, if I run just that one test, it will run in 20s rather than 99s. But it will make no difference to overall time to run CI, or on the likelihood that CI will time out, because we run all the tests together and they're already somewhat parallelized.

In other words, we have N CPUs to use. Making 1 test use N CPUs provides no benefit if those N CPUs are already in use by N tests.

**Comment by RaduBerinde** [comment]
I think that would be true in theory if the tests were completely pegged on CPU but I believe there is also waiting for conditions (in the tests and probably within the various subsystems as well). As a data point, on TestInitialPartition parallelized 8-way, I see less than 50% CPU usage on 4 CPUs.

**Comment by andy-kimball** [comment]
Hm, but then we're making our tests more complex and harder to debug, in the name of hopefully better utilizing CPU in an individual test. Wouldn't it make more sense to up the parallelization factor across tests so that we better utilize CPU more globally, if that's a problem?

**Comment by RaduBerinde** [comment]
It's also in the name of running tests for a single package faster (which happens frequently when someone is modifying a package, and is probably more important than CI runs).

I think we've had issues with tests timing out and had to reduce parallelism for some cases. Ideally it would be adaptive to what the tests are doing.

**Comment by andy-kimball** [comment]
But even in that case, I'd suggest that a better fix would be to break those really long-running tests into multiple tests, and then let regular Go test parallelization work. The problem with the gist you showed is that it makes that test harder to understand. I don't think we want to set a precedent where we expect intra-test parallelization. We should let the test framework do inter-test parallelization.

**Comment by andy-kimball** [comment]
+@danhhz in case he has other ideas on those long-running partitioning tests.

**Comment by RaduBerinde** [comment]
That's pretty much what I did, except I split them up into a few sub-tests instead of a few top-level tests. It's still the go test framework that decides when and how to run the parallel subtests.

**Comment by andy-kimball** [comment]
I know it's the same thing; I just have found intra-test parallelization to be difficult to understand when I come across it. I really have to stare hard at it to separate out the code doing parallelization from the code actually doing the testing. I try to only use it when I'm explicitly testing parallelization; not for the sole purpose of speeding up the test itself. But, if others don't feel similarly, I'm not dogmatic about that.

**Comment by danhhz** [comment]
I took a quick glance at the partitioning tests when I saw Radu's PR to see if there were some settings we could tweak to speed them up, but I didn't find anything obvious. Each of the sub tests does seem to take a second or two, though, which does seem too slow, so I haven't given up on finding something here

**Comment by hueypark** [comment]
@RaduBerinde  Could you update checklist for all function with priority(execution time will be best)? This will drive further work manageable.

**Comment by RaduBerinde** [comment]
The checklist has running times. Not sure what else you need.

**Comment by hueypark** [comment]
This comment(https://github.com/cockroachdb/cockroach/issues/37028#issuecomment-495063699) seems to have expanded the issue.  Is it better to create new issue?

**Comment by hueypark** [comment]
@RaduBerinde I created improved PR(#39142) for this issue. Could you review this?
